### PR TITLE
[MTP][GLM5.2] Remove copy and fuse eltwise in draft module at cg mode

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
## Summary
- Reuse fixed DraftGraph buffers for MTP token IDs and hidden states, eliminating redundant D2D staging copies.
- Fuse replicated embedding, dual RMSNorm, and per-token FP8 quantization into one Triton kernel.
- Preserve safe fallbacks for unsupported embedding and quantization configurations.

### CODE REVIEW

**Comment by ZhiweiYan-96** [comment]
# Trace proof

## Copy removal

* Two copy at start of propose_eagle, copy input_ids and hiddenstates

Baseline:
image

Opt:
image

**Save time: 8.638us**

* Prologue, copy for position at tails of draft step

Baseline:
image

Opt:
image

**Save time: 8.878us**


## Fusion for embedding,rmsnorm,quant

Baseline 

image

Opt

image

**Save time: 12.877us->4.959us**


## Conclusion

Overtime for 1 draft step: 464us->433us

image


image

File: atom/models/deepseek_mtp.py [hunk 1]
```diff
@@ -200,15 +209,36 @@ def forward(
...
+        eh_input_quant = None
+        can_fuse_prologue = (
+            inputs_embeds is None
+            and isinstance(self.embed_tokens, ReplicatedEmbedding)
+            and input_ids.ndim == 1
+            and previous_hidden_states.ndim == 2
+            and layer.eh_proj.quant_type.value == QuantType.per_Token.value
+            and layer.eh_proj.params_dtype == dtypes.fp8
+            and getattr(layer.eh_proj, "input_scale", None) is None
+        )
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/models/deepseek_mtp.py:230]: add fallback for unfused path

File: atom/spec_decode/draft_graph.py [hunk 2]
```diff
@@ -19,16 +19,19 @@
...
+    pool may re-issue its storage on a later replay.
     """
     if isinstance(value, torch.Tensor):
+        if any(value.data_ptr() == tensor.data_ptr() for tensor in staged):
+            return value
         return value.clone()
     if isinstance(value, tuple):
-        return tuple(_owned(v) for v in value)
+        return tuple(_owned(v, staged) for v in value)
     return value
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/spec_decode/draft_graph.py:35]: test cases added

File: atom/models/deepseek_mtp.py [hunk 3]
```diff
@@ -280,6 +312,10 @@ def compact_topk_indices(self, slot_ids: torch.Tensor) -> None:
 
 @support_torch_compile
 class DeepSeekMTP(nn.Module):
+    # EagleProposer may write the next-step ids/hidden directly into its fixed
+    # DraftGraph inputs. Keep this opt-in: other draft architectures can carry
+    # different hidden-state aliasing contracts.
+    supports_draft_buffer_reuse = True
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/models/deepseek_mtp.py]: fixing

[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/models/deepseek_mtp.py]: removed

File: atom/model_ops/fused_mtp_prologue.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    embed = tl.load(
+        embedding_weight_ptr + token_id * embedding_stride_row + cols,
+        mask=token_valid & mask,
+        other=0.0,
+    ).to(tl.float32)
+    hidden = tl.load(
+        hidden_ptr + row * hidden_stride_row + cols,
+        mask=mask,
+        other=0.0,
+    ).to(tl.float32)
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/fused_mtp_prologue.py]: fixed in the latest push. Address offsets now use int64, and invalid token IDs use a safe index before the masked load. I also added tests for int32/int64 IDs and both negative and positive out-of-range values.

File: atom/model_ops/attentions/aiter_mla.py [hunk 5]
```diff
@@ -66,6 +66,22 @@ def _mla_seg_meta_kwargs() -> dict:
...
+def _pad_prefill_mla_draft_tail(
+    kv_indptr: torch.Tensor,
+    kv_last_page_lens: np.ndarray,
+    block_tables: np.ndarray,
+    scheduled_bs: int,
+    running_bs: int,
+) -> None:
+    """Give widened Eagle rows an empty, fully initialized MLA KV range."""
+    assert 0 < scheduled_bs <= running_bs
+    if scheduled_bs == running_bs:
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply atom/model_ops/attentions/aiter_mla.py:78]: relaxed the assertion and added a regression test to verify all padded MLA metadata is safely zero-initialized for an empty rank.

**Comment by ZhiweiYan-96** [comment]
Accuracy result (numfew_shot=20)

image

**Review by valarLip — APPROVED** [review]
